provider/ec2/environ.go

```diff
@@ -744,6 +744,20 @@ func (e *environ) StartInstance(
 		e: e,
 		i: instResp.Instances[0],
 	}
+	// In the case the allocate-public-ip constraint is set to false, we
+	// need to dissociate the public IP address from the instance.
+	allocatePublicIP := args.Constraints.AllocatePublicIP
+	if allocatePublicIP != nil && !*allocatePublicIP {
+		// We can safely dissociate the public IP address from the
+		// first (primary) network interface of the instance, because
+		// (from the AWS docs):
+		// This option can be enabled for any network interface but will only apply to the
+		// primary network interface (eth0).
+		e.ec2Client.ModifyNetworkInterfaceAttribute(ctx, &ec2.ModifyNetworkInterfaceAttributeInput{
+			NetworkInterfaceId:       inst.i.NetworkInterfaces[0].NetworkInterfaceId,
+			AssociatePublicIpAddress: aws.Bool(false),
+		})
```

Comment:
Note to reviewers: I haven't been able to counter-proof that an instance could have 0 NICs by default, is this hypothesis correct?